tests/samples/simple_pack_pipeline_e2e.mlir

```diff
@@ -44,3 +44,31 @@ func.func @matmul_large(%lhs: tensor<2048x512xi32>, %rhs: tensor<512x2048xi32>)
                     outs(%fill: tensor<2048x2048xi32>) -> tensor<2048x2048xi32>
   return %res : tensor<2048x2048xi32>
 }
+
+// -----
+
+// This test demonstrates Pack pipeline based e2e lowering for matmul transpose.
+
+// CPP-LABEL: hal.executable.export public @matmul_transpose_static_dispatch_0_generic_8x32x16_i32
+//       CPP:    aie.device(ipu)
+//       CPP:    aie.shim_dma_allocation
+//       CPP:    aie.shim_dma_allocation
+//       CPP:    aie.shim_dma_allocation
+//       CPP:    func.func @matmul_transpose_static_dispatch_0_generic_8x32x16_i32(%arg0: memref<8x16xi32>, %arg1: memref<32x16xi32>, %arg2: memref<8x32xi32>)
+//       CPP:      aiex.ipu.dma_memcpy_nd
+//       CPP:      aiex.ipu.dma_memcpy_nd
+//       CPP:      aiex.ipu.dma_memcpy_nd
+//       CPP:      aiex.ipu.sync
+func.func @matmul_transpose_static(%lhs : tensor<8x16xi32>,
+    %rhs : tensor<32x16xi32>) -> tensor<8x32xi32> {
+  %cst = arith.constant 0 : i32
+  %empty = tensor.empty() : tensor<8x32xi32>
+  %fill = linalg.fill ins(%cst : i32) outs(%empty : tensor<8x32xi32>) -> tensor<8x32xi32>
+  %matmul_transpose = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d2)>, affine_map<(d0, d1, d2) -> (d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%lhs, %rhs : tensor<8x16xi32>, tensor<32x16xi32>) outs(%fill : tensor<8x32xi32>) {
```

Comment:
I think we want to support the named op, matmul.tranpose_b from the OPT dispatch tracker, note that it is still can be interacted as a generic in the kernelDispatch code?

Reply:
Maybe I'm missing something but I didn't see a `linalg.matmul_transpose_b` op in the OPT dispatch tracker's listed dispatches.

My first hunch was to raise `linalg.generic` -> `linalg.matmul_transpose_b` but I guess that can't be done because the `matmul_transpose_b` apparently has some elemental type constraint (@MaheshRavishankar please correct me if I'm wrong).

And if we go ahead with just `matmul_transpose_b`, the AffineMap checks I added would still be need to differentiate amongst the various ops implementing `ContractionOpInterface`. But that requires either the `linalg.matmul_transpose_b` to be directly present in the dispatch OR raising of `linalg.generic` (still using the matcher check) to `matmul_transpose_b` and then moving on to KernelDispatch's handle.

Reply:
oh you are right about them being generic in OPT dispactches, I got confused becuase the liner dispatch Kumar was showing the other day had transpose b named op. Anyway, in that case my only comment is to adapt the transpose logic from the CPU backend then coming up with a new one?

Reply:
So the [transpose check](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1499-L1523) in CPU is about a `linalg.generic` op which just performs a transpose. That check is not adaptable for that reason besides it being a `static` function defined just within that file.

Please let me know if there's any such utility in the CPU backend that I should look into and can be reused (copy/paste) in this case.

Reply:
Your check is also for generic so I was thinking just copy paste the code in our file?

Reply:
ah, I thought it was doing a matmul transpose my bad! so how is the CPU dealing with a generic that does a matmul trasnpose, does it become a named op? Is it becoming a named op for us when we see it in the kernel dispatch code? I am just trying to avoid new logic (for the sake of lowering maintenance burden) so want to handle this how cpu/gpu handle it as at this level its fairly similar other than the actual config attriute we add to it.

Reply:
I dont think the elemtent type is used in the raise-special-ops matcher
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp#L75-L85
Anyway, its good to know this IR is going to default config for CPU, lets discuss in a call today to make progress on this. Thanks for investigating.

Reply:
I think that matcher is to match linalg.matmul(a, linalg.transpose(b)), not for raising from a generic, I dont think that raising exists right now.

Reply:
Yes, you're correct.
I did check that pass and suggested to add the `linalg.generic` -> `linalg.matmul_transpose_a/b` to upstream IREE earlier, but seems like certain element types aren't supported via `linalg.matmul_transpose_a/b` later down the stack. And maybe that, in a way, explains why CPU pipeline chooses to multiplex this to a default config and doesn't have this structure handled in `raise-special-ops`.

I cannot recall the exact reason though - one way would be to indeed have an IR starting from `linalg.matmul_transpose_b` and check how is that handled in the stack. If it remains as a `linalg.matmul_transpose_b` for the element type the test case is trying to handle - then perhaps it'd indeed make sense to use the matcher I added here to raise `linalg.generic` to `linalg.matmul_transpose_a/b`. 

Depending on what we converge on I can work on either of the following :-
1. Update `raise-special-ops`.
2. Keep the current implementation as it is.
3. A default config which would work on all non-transpose and non-element-wise should be thought of and added. I'll have to check what CPU is doing and understand that - but seems like this would open other can of worms because the current pipeline is fixated on certain loop tiling structure.

Reply:
It is worth checking what the CPU does for the tranpose_b named op, I think it wont apply the default config in that case, as it will treat it like a contracton op?
https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L1175

what is your reference for element type is the reason for it not getting raised? I cant see that anywhere.

Also, I am fine with landing this if its not getting raised, however, one thing to handle is what if the dispatch did actually present as mamtul.transpose_b, I saw that in a simple linear layer example Kumar had, and could happen if changes happen in the front end, so dont want to break in that case. However, that can be a separate PR. Let me know what you think.

Reply:
> what is your reference for element type is the reason for it not getting raised? I cant see that anywhere.

I was citing that based on an earlier discussion I've had with @MaheshRavishankar when I was talking about the `linalg.generic` -> `linalg.matmul_transpose_a/b` - which I've verified today (see below).

> one thing to handle is what if the dispatch did actually present as mamtul.transpose_b

Yes, absolutely. I did try that in an IR that itself starts with `matmul_transpose_b` to check how CPU is handling that - hopefully now it gives us a complete overview of what's happening at KernelDispatch level - it indeed goes through ContractionOp's case. I see that there's no issue in the lowering with the element type `i32` and I verified the results during the execution of the generated vmfb too. `linalg.matmul_transpose_a/b` goes up until Vectorization phase, post which it vanishes because of vectorization for `i32`'s case.

For `f32` it gets converted to a `linalg.mmt4d` op. Results were verified here too.

You're correct about the front-end aspect - But adding support for `linalg.matmul_transpose_a/b` too right now wouldn't be too fruitful since the required element type's handling is `f32` - and it is indeed not supported by both `linalg.matmul_transpose_a/b` as well as the current AMD-AIE `linalg.matmul` itself is unable to handle `f32`.

So, we've got a lot of iterations to handle these incrementally.

Reply:
About the last point, I think the element type support and shape support are orthogonal, but so is supporting this type of generic matmul traponse and matmul.tranpose_b op, so its fine to do this and add the tranpose_b later. However, I am not sure if Mahesh's earlier comment about using the contraction op interface is actually being done here, so I would check with him on that.